### ISSUE TITLE
Oracle "Maximum Open Cursors Exceeded" Error During Sync Process

### ISSUE THREAD

**Issue by orware** [metabase/metabase#4389]
### Bugs

On my end below it may be a configuration thing (I've contacted the person that manages this database here on campus to see if they can do some checking for me), but I've connected two different Oracle databases to Metabase. In the first case I didn't run into the error below, but in the second case (which is a different database on a different server so the config may be different) it seems the sync process runs into the issue below. What I'm not sure about is how many connections Metabase is trying to open up to handle the syncing process?

It seems like it should be able to try and handle the syncing process more slowly in a situation like the one below and perhaps throttle back the number of connections it will use and retry (if the number of cursors are potentially limited). Obviously having more cursors available would be better, which is why I reached out to my contact here on campus, but just thought I'd share it with you because it may be something that others run into as well. At the moment though it seems that this particular issue is going to prevent me from using the additional Oracle database for at least a while since it can't currently sync with it properly.

```
02-15 10:09:28 [1mERROR task.sync-databases[0m :: Error syncing database 3:  #error {
 :cause ORA-00604: error occurred at recursive SQL level 1
ORA-01000: maximum open cursors exceeded
ORA-00604: error occurred at recursive SQL level 1
ORA-01000: maximum open cursors exceeded
ORA-01000: maximum open cursors exceeded

 :via
 [{:type java.sql.SQLException
   :message ORA-00604: error occurred at recursive SQL level 1
ORA-01000: maximum open cursors exceeded
ORA-00604: error occurred at recursive SQL level 1
ORA-01000: maximum open cursors exceeded
ORA-01000: maximum open cursors exceeded
```

**Comment by camsaul** [comment]
@orware for SQL databases, Metabase sets up connection pools with a max of 15 connections

**Comment by camsaul** [comment]
It sounds like this DB is configured with a custom connection limit set. Do you know what it is?

**Comment by cletisgipson** [comment]
I am experiencing this exact same issue with my oracle servers.  I had the oracle dba increase the number of allowable open cursors, this worked temporarily.  But it the issue has came back.

**Comment by ricardoj-br** [comment]
I'm having the same issue when connecting to Oracle. Is there any workaround besides asking DBA to increase the number of cursors ? Thanks.

**Comment by MarkRx** [comment]
See comment on https://github.com/metabase/metabase/issues/6467#issuecomment-361972742. In oracle there can be multiple cursors per a connection. If more cursors are opened before others are closed (possibly due to recursion) you can run out of cursors. It's a bug for sure.

**Comment by senior** [comment]
This issue appears to have been resolved by ensuring the ResultSets returned by the JDBC metadata calls get closed explicitly, which was added in https://github.com/metabase/metabase/pull/7666. If you're still hitting this issue with version `0.30.3`+, ensure you are using the latest Oracle JDBC Driver (`12.2.0.1.0` and `ojdbc8.jar`).

**Comment by orware** [comment]
@senior I've updated to 0.30.4 (I was already on 0.30.3) and updated to the newer ojdbc8.jar file (on the Windows Server I'm running Metabase on, and just grabbed it from this page on the Oracle site: https://www.oracle.com/technetwork/database/features/jdbc/jdbc-ucp-122-3110062.html).

I ran a quick query to make sure the Oracle connectivity was still working ok and it seemed to be, so hopefully this helps with this issue, though I've had our setup on a manual sync for a while in Metabase to help avoid the issue.

Up until this point we've been running on the ojdbc6.jar file.
